### DISCUSSION TITLE
Build Metrics

### DISCUSSION THREAD

**Discussion by zachgrayio** [MobileNativeFoundation/discussions#4] [Proposals]
Since I know this means a lot of different things to folks, it's probably best to start a thread to just collect input for awhile before spending much time drafting any specific proposals. 

Let's kick it off!

**Comment by keith** [comment]
Can you clarify a bit what the scope of this question is? I know there's a lot of threads with the mocks that are floating around, but are you most interested in hearing the data points folks care about or something else?

**Reply by zachgrayio** [reply]
Yes!

I think definitely we'll want to discuss specific data to be captured in great detail, but a few other questions before that even:

1: how do folks envision the actual implementation of 1st the gathering and then sharing of these metrics?

There's already some prior art on gathering metrics (xcmetrics, Flank's new metrics project, various Bazel-based projects and products, gradle enterprise, etc); unfortunately all very fragmented due to different schemas and specificity to certain tools.

_I for one would think that standardization of a common API (maybe a .proto) might be useful here; but I also know that some folks have been happy with a poll and/or spreadsheet in the past, so really just looking for input and opinions before going too far._ 

2: if a metrics protocol is defined, should the MNF host a shared backend server implementing it to facilitate the sharing of data? if such a central metrics repository were to be maintained, what access levels are appropriate? MNF members only? public? member-to-member sharing only? a combination of all of these?

3: what kind of user data redaction is necessary? what sensitive metrics, if any, should be captured for internal/private use but redacted from sharing? should the metrics system meet both internal-only and shared metrics use-cases?

There's more, but this is probably enough to start with.

**Reply by dinesh-hd** [reply]
Here is a common conversation between a feature engineer and a build engineer.

**Feature Engineer to build engineer** : My builds are slow. Can you fix it?

**Build Engineer to himself :** 

- What is causing the slow down?
- What are the components of the build graph?
- Which component in the build graph is taking the most time to build?
- What is the development environment where the build is slow? What is the RAM? What is the CPU? What is the IDE setting? Is it a clean build or an incremental build?
- Is there a specific commit that has caused the slow down ? [Commit to build metric mapping]

A framework to collect and answer these metric would be very benificial. 

PS : _I may be a bit greedy. But I am so in for contributing._

**Comment by shepting** [comment]
We would be quite curious to know how our build times stack up to other companies. Perhaps some potential value for sharing would be for recruiting reasons.

We'd ideally like to see it split up between full builds with `xcodebuild`, builds from the standard IDE (Xcode) with a more "modern" build system like Buck/Bazel, and then builds of a subset of the app (like dev/sandbox/playground) types of builds that most companies have. I'm guessing we'd want to share p50 and p90/p95 at the very least and likely also p75.

**Comment by gtroshin** [comment]
I am not sure if this discussion for the iOS particularly, but in Android world the best tool that is widely used (as far as I know) is Gradle, Gradle Enterprise and Gradle Scan. These are best tools to monitor the build metrics and how it improves or degrades.

**Reply by zachgrayio** [reply]
Very important to capture android as well, so touching Gradle (and bazel/buck) for sure. Definitely the gold standard, but we also need portability

**Comment by dinesh-hd** [comment]
There are 2 major dimensions that would be of use
- Build metrics for a whole team
- Build metric per build

The parameters that would be of use are as follows

Having seen both Android and iOS builds, there are some commonalities that would greatly help improve developer experience
- Build cache hit/miss ratio [Applies to both Android and iOS] - This can give a fair idea of how the avoidance savings and how builds can be improved
- Incremental build time / clean build time 
- Time taken per unit of build : Time taken to build a specific module. This can give the distribution of build times per build
- Critical path for a build target  : The total number fo sequential dependencies that occur in the build path
- Environment metrics : Machine type, available ram, available CPU, CPU usage, thread counts, JVM version, XCODE/Android Studio versions.

I do agree that Gradle Enterprise is a good option but it takes lot of cost to set it up and operate at scale. 
An open source alternative would be good.

Talaoit for Android does a good Job on metric collection on Gradle Builds.

**Reply by zachgrayio** [reply]
Yeah, `Talaoit` looks like the ideal integration point and I have some early work on this already for some of our internal stuff

**Comment by BalestraPatrick** [comment]
As one of the developers behind XCMetrics (http://github.com/spotify/XCMetrics/), I'm very much interested in this topic. One challenge that I see in coming up with a possible set of shared metrics to be collected while allowing consumers to pick and choose which metrics they'd like to store. For example, XCMetrics was modelled after the Xcode `xcactivitylog` file format and the information contained in it, and it has been augmented with other information that plugins can collect. This allows consumers to write their own plugins to collect even more information such as git data, RAM usage, or project setup metadata. I think this is very important to keep in mind since everyone's requirements differ.

I also think this type of service should definitely be self-hosted for the most part. Trying to come up with a central service to collect metrics would quickly encounter scaling problems as well as privacy and security concerns (build logs usually contain a lot of unreleased features which shouldn't be made public). With XCMetrics we tried to make it as easy as possible to get started by providing ready-to-be-deployed Docker images as well as Kubernetes configuration files.

**Reply by dinesh-hd** [reply]
Thats a really nice approach. Loved the Docker based set up. So easy to set up and get rolling.

**Reply by zachgrayio** [reply]
Yeah I agree on the hosting. I think we should consider providing sort of mechanism to export/publish/share redacted data for comparison somehow as well though as that seems to be a big driver for some folks. An analogy to prometheus comes to mind, where each private deployment of the service might expose a `metrics` endpoint to be scraped, or use generated gRPC clients to push to some repo, etc.

**Reply by tinder-maxwellelliott** [reply]
Wow this is really impressive @BalestraPatrick ! I just learned about this today.

Ideally we could somehow add a plugin to this repo that would allow us to export this data in a safe way and then it could be shared here. I think if companies could compare their p90s, p50s and cache hit ratios it takes it out of the hypothetical when pushing for improvements internally.

**Reply by zachgrayio** [reply]
Yea XCMetrics is definitely great; the challenge is that its just for xcode and we want to capture the breadth of what MNF folks are doing so at a minimum
- gradle
- bazel
- buck
- xcodebuild

to that end I've been actually working on a fork of xcmetrics (it is public, though not ready to spend much time looking at yet) which sends the data to a bazel BES backend (we have done this for a few other tools) but if I'm honest this is pretty hacky and a big motivation for us on a metrics standard is that this really doesn't feel like the right protocol either, it's just the one we've operationalized for customers so its where we've been going so far. 

but @BalestraPatrick and team have really done a great job in mapping out the space here so it could be the case that this project is a good jumping off point, im fairly agnostic (other than being a big fan of the work they have done!)

i havent grokked the plugins part yet, sounds like i need to take a closer look there.

**Reply by BalestraPatrick** [reply]
I think the biggest challenge in building a single tool to capture metrics across all the build systems is that most of them have different models for collects metrics. Bazel and Buck model builds after actions and start/finish events ([proto example](https://github.com/bazelbuild/bazel/blob/8f927d2e612263c78f6f4122f634434c29c71c85/src/main/java/com/google/devtools/build/lib/buildeventstream/proto/build_event_stream.proto#L942-L972)), whereas other build systems may model it differently via steps/targets. Basically it'll be hard to find a 1:1 direct mapping between all the build systems and a common metrics collection API. This is not impossible to solve, but it could lead to some kind of data incompleteness. I don't think it's such a big deal, but something to keep in mind.

**Reply by Viniciuscarvalho** [reply]
In the case of iOS, did any of you use Apple's own MetricKit on a large scale?
